lib/std/smartptrs.nim

```diff
@@ -0,0 +1,139 @@
+  p.val[]
+
+proc `$`*[T](p: UniquePtr[T]): string {.inline.} =
+  if p.val == nil: "UniquePtr[" & $T & "](nil)"
+  else: "UniquePtr[" & $T & "](" & $p.val[] & ")"
+
+#------------------------------------------------------------------------------
+
+type
+  SharedPtr*[T] = object
+    ## Shared ownership reference counting pointer.
+    val: ptr tuple[value: T, atomicCounter: int]
+
+proc `=destroy`*[T](p: var SharedPtr[T]) =
+  if p.val != nil:
+    if (when compileOption("threads"):
+          atomicLoadN(addr p.val[].atomicCounter, ATOMIC_CONSUME) == 0 else:
+          p.val[].atomicCounter == 0):
+      `=destroy`(p.val[])
+      when compileOption("threads"):
+        deallocShared(p.val)
+      else:
+        dealloc(p.val)
+    else:
+      when compileOption("threads"):
+        discard atomicDec(p.val[].atomicCounter)
+      else:
+        dec(p.val[].atomicCounter)
+
+proc `=`*[T](dest: var SharedPtr[T], src: SharedPtr[T]) =
+  if src.val != nil:
+    when compileOption("threads"):
+      discard atomicInc(src.val[].atomicCounter)
+    else:
+      inc(src.val[].atomicCounter)
+  if dest.val != nil:
+    `=destroy`(dest)
+  dest.val = src.val
+
+proc newSharedPtr*[T](val: sink Isolated[T]): SharedPtr[T] {.nodestroy.} =
+  when compileOption("threads"):
+    result.val = cast[typeof(result.val)](allocShared(sizeof(result.val[])))
+  else:
+    result.val = cast[typeof(result.val)](alloc(sizeof(result.val[])))
+  result.val.atomicCounter = 0
+  result.val.value = val.extract
+
+converter convertSharedPtrToObj*[T](p: SharedPtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    doAssert(p.val != nil, "deferencing nil shared pointer")
+  p.val.value
+
+proc isNil*[T](p: SharedPtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc `[]`*[T](p: SharedPtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    doAssert(p.val != nil, "deferencing nil shared pointer")
+  p.val.value
+
+proc `$`*[T](p: SharedPtr[T]): string {.inline.} =
+  if p.val == nil: "SharedPtr[" & $T & "](nil)"
+  else: "SharedPtr[" & $T & "](" & $p.val.value & ")"
+
+#------------------------------------------------------------------------------
+
+type
+  ConstPtr*[T] = distinct SharedPtr[T]
+    ## Distinct version of referencing counting smart pointer SharedPtr[T],
+    ## which doesn't allow mutating underlying object.
+
+proc newConstPtr*[T](val: sink Isolated[T]): ConstPtr[T] =
+  ConstPtr[T](newSharedPtr(val))
+
+converter convertConstPtrToObj*[T](p: ConstPtr[T]): lent T {.inline.} =
+  SharedPtr[T](p).val.value
+
+proc isNil*[T](p: ConstPtr[T]): bool {.inline.} =
+  SharedPtr[T](p).val == nil
+
+proc `[]`*[T](p: ConstPtr[T]): lent T {.inline.} =
+  when compileOption("boundChecks"):
+    doAssert(SharedPtr[T](p).val != nil, "deferencing nil const pointer")
+  SharedPtr[T](p).val.value
+
+proc `$`*[T](p: ConstPtr[T]): string {.inline.} =
+  if SharedPtr[T](p).val == nil: "ConstPtr[" & $T & "](nil)"
```

Comment:
showing the **static** type is redundant, `$` in nim almost never shows it, eg:
```nim
type Foo =  object
  x: int
assert $Foo(x: 1) == "(x: 1)"
```
and for good reasons (eg you'd end up otherwise with lots of redundant names inside a `seq[Foo]`


to show more debug info (including type), we'll instead use std/prettyprint, see proposal 2 in https://github.com/nim-lang/RFCs/issues/203#issuecomment-602534906 (i already implemented it in private branch)

Reply:
make sense

fixed